### PR TITLE
REGR: allow reindexing datetimelike with upcast / raise deprecation warning

### LINKED ISSUE #42921: BUG: Reindex null timedelta series with `pd.Timedelta` fails
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
ts = pd.date_range("2020-01-01", "2020-01-03", freq="1d")
s = pd.Series(pd.Timedelta(None), index=ts)
s.reindex(pd.date_range("2020-01-01", "2020-01-04", freq="1d"), fill_value=pd.Timedelta(0))
```

#### Problem description

To begin with, `s` has the wrong type `datetime64[ns]` instead of `timedelta64[ns]`. This causes the `reindex` to fail.
I've also tested the code above on version 1.2.4, it does not produce error and the type is casted as `object`:
image

#### Expected Output
image

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : c7f7443c1bad8262358114d5e88cd9c8a308e8aa
python           : 3.7.11.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : None.None

pandas           : 1.3.1
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.2
setuptools       : 52.0.0.post20210125
Cython           : 0.29.24
pytest           : 6.1.2
hypothesis       : 6.14.1
sphinx           : 4.0.2
blosc            : None
feather          : None
xlsxwriter       : 1.4.4
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : 1.0.2
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.22.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 2021.07.0
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.2
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : 0.4.2
scipy            : 1.6.2
sqlalchemy       : 1.4.22
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.19.0
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.0

</details>

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes https://github.com/pandas-dev/pandas/issues/42921

- Fixes inconsistency/regression in the deprecation of `date` coercion (https://github.com/pandas-dev/pandas/pull/39767) -> now this is also done for single-axis reindexing (xref https://github.com/pandas-dev/pandas/pull/44798#issuecomment-988264205)
- This also fixes the regression reported in https://github.com/pandas-dev/pandas/issues/42921(although it is not fully clear if we really want to "fix" this regression)

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
Thanks @jorisvandenbossche 

just to be clear...

```python
s.reindex(
    pd.date_range("2020-01-01", "2020-01-04", freq="1d"), fill_value=pd.Timedelta(0)
)
```

on 1.2.5

```
2020-01-01               None
2020-01-02               None
2020-01-03               None
2020-01-04    0 days 00:00:00
Freq: D, dtype: object
```

and with this PR
```
2020-01-01                NaT
2020-01-02                NaT
2020-01-03                NaT
2020-01-04    0 days 00:00:00
Freq: D, dtype: object
```

so this is not restoring 1.2.5 behaviour exactly.

However, the dtype is still `object`, whereas in the issue OP the expected behavior is for `timedelta64[ns]` (which is the result when dtype is explictly specfied, second part of https://github.com/pandas-dev/pandas/issues/42921#issuecomment-897568579)

so maybe should not yet close #42921 with this PR?

**Comment by jreback** [comment]
moving this to 1.4. we don't need to backport this

**Comment by jorisvandenbossche** [comment]
I moved this back to 1.3.5 milestone for now (let's discuss first before moving). 

It _does_ fix a regression from 1.2->1.3. Code that worked on 1.2, is raising an error on 1.3. 

@simonjayhawkins you're right that None vs NaT is not exactly the same (I didn't notice that yet, can look into why that is the case), but it's also a much smaller difference compared to raising an error in 1.3


> However, the dtype is still `object`, whereas in the issue OP the expected behavior is for `timedelta64[ns]` (which is the result when dtype is explictly specfied, second part of [#42921 (comment)](https://github.com/pandas-dev/pandas/issues/42921#issuecomment-897568579))

As you explained in that issue, it is expected that this doesn't give `timedelta64[ns]`, since the example starts with a datetime64 dtype.

**Comment by simonjayhawkins** [comment]
> As you explained in that issue, it is expected that this doesn't give `timedelta64[ns]`, since the example starts with a datetime64 dtype.

That's fine. we probably need to repeat/confirm this in the issue (that the expected in the OP is incorrect) so that it can be closed off.

**Comment by jreback** [comment]
@jorisvandenbossche it may be a regression but this is delaying the long delayed release

**Comment by jreback** [comment]
it's simply too late

**Comment by jreback** [comment]
@jorisvandenbossche pls don't move this back
it is too late for this patch which hasn't been looked at and will not before release

**Comment by jbrockmendel** [comment]
will cases that go through ExtensionBlock.take_nd be affected/corrected?

File: doc/source/whatsnew/v1.3.5.rst [hunk 1]
```diff
@@ -21,6 +21,7 @@ Fixed regressions
 - Fixed performance regression in :func:`read_csv` (:issue:`44106`)
 - Fixed regression in :meth:`Series.duplicated` and :meth:`Series.drop_duplicates` when Series has :class:`Categorical` dtype with boolean categories (:issue:`44351`)
 - Fixed regression in :meth:`.GroupBy.sum` with ``timedelta64[ns]`` dtype containing ``NaT`` failing to treat that value as NA (:issue:`42659`)
+- Fixed regression in :meth:`~Series.reindex` raising an error when using an incompatible fill value with a datetime-like dtype (or not raising a deprecation warning for using a ``datetime.date`` as fill value) (:issue:`42921`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.5.rst:24]: move to 1.4

File: pandas/tests/frame/methods/test_reindex.py [hunk 2]
```diff
@@ -113,6 +116,11 @@ def test_reindex_date_fill_value(self):
         )
         tm.assert_frame_equal(res, expected)
 
+        # only reindexing rows
+        with tm.assert_produces_warning(FutureWarning):
```
Comment (jreback) [inline pandas/tests/frame/methods/test_reindex.py:120]: why is this a warn?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_reindex.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/tests/frame/methods/test_reindex.py:120]: xref https://github.com/pandas-dev/pandas/pull/44798#issuecomment-988264205.

@jorisvandenbossche maybe open dedicated issue for discussion/visibility.

Comment (jorisvandenbossche) [reply pandas/tests/frame/methods/test_reindex.py:120]: I think this PR can serve that purpose?

**Comment by jorisvandenbossche** [comment]
> will cases that go through ExtensionBlock.take_nd be affected/corrected?

No, because they are already strict in preserving the dtype anyway

**Comment by jbrockmendel** [comment]
>> will cases that go through ExtensionBlock.take_nd be affected/corrected?

> No, because they are already strict in preserving the dtype anyway

It isn't clear to me that this strictness is correct.  i.e. in test_reindex_fill_value_datetimelike_upcast what happens if you add a couple of EA dtypes to the 'dtype' param?

**Comment by jorisvandenbossche** [comment]
> It isn't clear to me that this strictness is correct.

It has been like that for a long time (for ever?), and changing (or discussing) that is out of scope for this PR I think.

**Comment by jbrockmendel** [comment]
> that is out of scope for this PR I think.

Sure.

Looks like ArrayManager tests are failing.

File: pandas/core/array_algos/take.py [hunk 3]
```diff
@@ -94,6 +94,12 @@ def take_nd(
     """
     if fill_value is lib.no_default:
         fill_value = na_value_for_dtype(arr.dtype, compat=False)
+    elif isinstance(arr.dtype, np.dtype) and arr.dtype.kind in "mM":
```
Comment (jbrockmendel) [inline pandas/core/array_algos/take.py:97]: does take_1d need to be changed to match?

**Comment by jbrockmendel** [comment]
@jorisvandenbossche needs rebase

**Comment by jreback** [comment]
@phofl @jbrockmendel if one of you could rebase this

**Comment by phofl** [comment]
rebased

**Comment by jreback** [comment]
@jbrockmendel ok here?

File: pandas/tests/frame/methods/test_reindex.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/frame/methods/test_reindex.py:120]: could use a `match=...` to clarify

**Comment by jbrockmendel** [comment]
Nothing that should block this.  I'd like to get a response here https://github.com/pandas-dev/pandas/pull/44839/files#r774813668 eventually

**Comment by phofl** [comment]
The array_manager runs through take_1d which does not promote to object in case of incompatible fill values. Not familiar enough with the logic to decide on a fix here

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
> The array_manager runs through take_1d which does not promote to object in case of incompatible fill values. Not familiar enough with the logic to decide on a fix here

Yah, trying to make AM behavior consistent is a fool's errand.

**Comment by jreback** [comment]
@jbrockmendel or @phofl can you rebase this

**Comment by jreback** [comment]
@jbrockmendel or @phofl if you can rebase this

**Comment by jbrockmendel** [comment]
merged master

**Comment by jreback** [comment]
hmm a code check / pre-commit issue if you can fixup @jbrockmendel

**Comment by jreback** [comment]
@meeseeksdev backport 1.4.x

**Pull request merged**